_posts/2016-12-17-writing-efficient-numerical-code.md

```diff
@@ -0,0 +1,347 @@
+| bilateralFilter (5x5)|16718651|4597027|263 |
+|------------------------------------------------------------------------------------------|
+| calcGradients|2244798|506101|343 |
+|------------------------------------------------------------------------------------------|
+| calcPartialDerivatives|428318|141520|202 |
+|------------------------------------------------------------------------------------------|
+| canny|4108987|758240|441 |
+|------------------------------------------------------------------------------------------|
+| filterNonMaximum|477543|38968|1125 |
+|------------------------------------------------------------------------------------------|
+| nonMaximaSupression|588455|84436|596 |
+|------------------------------------------------------------------------------------------|
+| remap|225780|62089|263 |
+|------------------------------------------------------------------------------------------|
+| warp|235169|63821|268 |
+|------------------------------------------------------------------------------------------|
+
+As you can see, speedups are massive - average in this set is 1356.54%, or if written as multiplier:
+`mean(previous / current) = 14.57x`. But as I'll show you in this post, changes made to the algorithm implementations
+were trivial. If you'd like to see complete benchmarking results, please take a look at
+[PR](https://github.com/libmir/dcv/pull/58) implementing these changes.
+
+*Discalmer: Please keep in mind DCV project is still far too young to be compared against proven computer vision toolkits such as
+OpenCV. Optimizations done here are showing power of `mir.ndslice.algorithm`, but if you dive into implementation of these algorithms,
+you'll notice most of them are implemented naively, without extensive optimization techniques applied. In future
+we'll focus on [separable filtering](https://github.com/libmir/dcv/issues/85), followed by cache locality improvement.*
+
+--------------------------------------------------------------------------------------------
+
+## Examples
+
+I'd like to show few examples of `mir.ndslice.algorithm`, but first we'll take a
+look at the basic principle how it can efficiently replace loop-based code. And later
+on we'll see how it can be used in a bit more complex algorithms.
+
+### Basics
+
+So, let's first examine the basic principle of utilizing iteration algorithms. This principle is also the
+basis of that *DCV* refactoring I've mentioned. Say we have following code, written plainly in C-style
+loops:
+
+```d
+Slice!(2, float*) image; // initialized elsewhere...
+
+for(size_t r; r < image.length!0; ++r) {
+    for(size_t c; c < image.length!1; ++c) {
+        // perform some processing
+    }
+}
+```
+
+... This code can be rewritten like so:
+
+```d
+import mir.ndslice.algorithm : ndEach;
+
+image.ndEach!( (ref v)
+{
+    /* perform that processing from inside those loops */
+}, Yes.vectorized, Yes.fastmath );
```

Comment:
`Yes.fastmath` is useless for lambdas. A function with @fasmath should be defined.